README.md
# Terraform Provider for the Hetzner DNS

Terraform provider for managing zones and records in the [Hetzner DNS service](https://dns.hetzner.com/).

## Installation

### terraform 0.13+

Add into your Terraform configuration this code:

```hcl-terraform
terraform {
  required_providers {
    hdns = {
      source = "alxrem/hdns"
    }
  }
}
```

and run `terraform init`

### terraform 0.12 and earlier

1. Download archive with the latest version of provider for your operating system from
   [Github releases page](https://github.com/alxrem/terraform-provider-hdns/releases).
2. Unpack provider to `$HOME/.terraform.d/plugins`, i.e.
   ```
   unzip terraform-provider-hdns_X.Y.Z_linux_amd64.zip terraform-provider-hdns_* -d $HOME/.terraform.d/plugins/
   ```
3. Init your terraform project
   ```
   terraform init
   ```

## Usage

Read the [documentation on Terraform Registry site](https://registry.terraform.io/providers/alxrem/hdns/latest/docs).


CHANGELOG.md
## 0.3.0 (Unreleased)

## 0.2.1 (December 9, 2021)

NOTES:

* Updated README

## 0.2.0 (July 5, 2020)

IMPROVEMENTS:

* Add data source for zone
* Add importers

BUG FIXES:

* Avoid concurrent zone updates

## 0.1.0 (May 5, 2020)

NOTES:

First public release.


hdns/provider_test.go
package hdns

import (
	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
	"os"
	"testing"

	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
	"github.com/hashicorp/terraform-plugin-sdk/terraform"
)

var testAccProviders map[string]terraform.ResourceProvider
var testAccProvider *schema.Provider

func init() {
	testAccProvider = Provider().(*schema.Provider)
	testAccProviders = map[string]terraform.ResourceProvider{
		"hdns": testAccProvider,
	}
}

func TestMain(m *testing.M) {
	resource.TestMain(m)
}

func TestProvider(t *testing.T) {
	if err := Provider().(*schema.Provider).InternalValidate(); err != nil {
		t.Fatalf("err: %s", err)
	}
}

func TestProvider_impl(t *testing.T) {
	var _ terraform.ResourceProvider = Provider()
}

func testAccHdnsPreCheck(t *testing.T) {
	if v := os.Getenv("HDNS_TOKEN"); v == "" {
		t.Fatal("HDNS_TOKEN must be set for acceptance tests")
	}
}


hdns/resource_hdns_zone_test.go
package hdns

import (
	"context"
	"fmt"
	"github.com/alxrem/hdns-go/hdns"
	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
	"github.com/hashicorp/terraform-plugin-sdk/terraform"
	"log"
	"os"
	"strings"
	"testing"
)

const testAccPrefix = "hdns-testacc"

func testAccZoneName(rName string) string {
	return fmt.Sprintf("%s.dev", rName)
}

func init() {
	resource.AddTestSweepers("hdns_zone", &resource.Sweeper{
		Name: "hdns_zone",
		F:    testSweepZones,
	})
}

func TestAccHdnsZone_Basic(t *testing.T) {
	var zone hdns.Zone
	rName := acctest.RandomWithPrefix("hdns-testacc")

	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccHdnsPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccHdnsCheckZoneDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccHdnsCheckZoneConfig_basic(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccHdnsCheckZoneExists("hdns_zone.foobar", &zone),
					resource.TestCheckResourceAttr(
						"hdns_zone.foobar", "name", testAccZoneName(rName)),
					resource.TestCheckResourceAttr(
						"hdns_zone.foobar", "ttl", "900"),
				),
			},
			{
				Config: testAccHdnsCheckZoneConfig_changedTtl(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccHdnsCheckZoneExists("hdns_zone.foobar", &zone),
					resource.TestCheckResourceAttr(
						"hdns_zone.foobar", "name", testAccZoneName(rName)),
					resource.TestCheckResourceAttr(
						"hdns_zone.foobar", "ttl", "1800"),
				),
			},
			{
				ResourceName:      "hdns_zone.foobar",
				ImportState:       true,
				ImportStateVerify: true,
			},
		},
	})
}

//noinspection GoSnakeCaseUsage
func testAccHdnsCheckZoneConfig_basic(rName string) string {
	return fmt.Sprintf(`
resource "hdns_zone" "foobar" {
  name = "%s"
  ttl  = 900
}
`, testAccZoneName(rName))
}

//noinspection GoSnakeCaseUsage
func testAccHdnsCheckZoneConfig_changedTtl(rName string) string {
	return fmt.Sprintf(`
resource "hdns_zone" "foobar" {
  name = "%s"
  ttl  = 1800
}
`, testAccZoneName(rName))
}

func testAccHdnsCheckZoneDestroy(s *terraform.State) error {
	config := testAccProvider.Meta().(*providerConfig)
	client := config.client

	for _, rs := range s.RootModule().Resources {
		if rs.Type != "hdns_zone" {
			continue
		}

		zone, r, err := client.Zone.GetByID(context.Background(), rs.Primary.ID)
		if err != nil && r != nil && r.StatusCode != 404 {
			return fmt.Errorf(
				"error checking if Zone (%s) is deleted: %v",
				rs.Primary.ID, err)
		}
		if zone != nil {
			return fmt.Errorf("zone (%s) has not been deleted", rs.Primary.ID)
		}
	}

	return nil
}

func testAccHdnsCheckZoneExists(z string, zone *hdns.Zone) resource.TestCheckFunc {
	return func(s *terraform.State) error {
		rs, ok := s.RootModule().Resources[z]

		if !ok {
			return fmt.Errorf("not found: %s", z)
		}

		if rs.Primary.ID == "" {
			return fmt.Errorf("no Record ID is set")
		}

		config := testAccProvider.Meta().(*providerConfig)
		client := config.client

		// Try to find the key
		foundZone, _, err := client.Zone.GetByID(context.Background(), rs.Primary.ID)
		if err != nil {
			return err
		}

		if foundZone == nil {
			return fmt.Errorf("record not found")
		}

		*zone = *foundZone
		return nil
	}
}

func testSweepZones(_ string) error {
	client, err := createClient()
	if err != nil {
		return err
	}

	ctx := context.Background()
	zones, err := client.Zone.All(ctx)
	if err != nil {
		return err
	}
	log.Printf("[DEBUG] Found %d zones to sweep", len(zones))

	for _, z := range zones {
		if strings.HasPrefix(z.Name, testAccPrefix) {
			log.Printf("Deleting zone %s", z.Name)
			if _, err := client.Zone.Delete(ctx, z.ID); err != nil {
				return err
			}
		}
	}
	return nil
}

func createClient() (*hdns.Client, error) {
	if os.Getenv("HDNS_TOKEN") == "" {
		return nil, fmt.Errorf("empty HDNS_TOKEN")
	}
	opts := []hdns.ClientOption{
		hdns.WithToken(os.Getenv("HDNS_TOKEN")),
	}
	return hdns.NewClient(opts...), nil
}


hdns/resource_hdns_record_test.go
package hdns

import (
	"context"
	"fmt"
	"github.com/alxrem/hdns-go/hdns"
	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
	"github.com/hashicorp/terraform-plugin-sdk/terraform"
	"testing"
)

func TestAccHdnsRecord_Create(t *testing.T) {
	var record hdns.Record
	rName := acctest.RandomWithPrefix("hdns-testacc")

	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccHdnsPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccHdnsCheckRecordDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccHdnsCheckRecordConfig_create(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccHdnsCheckRecordExists("hdns_record.bar", &record),
					resource.TestCheckResourceAttr("hdns_record.bar", "name", "www"),
					resource.TestCheckResourceAttr("hdns_record.bar", "ttl", "86400"),
					resource.TestCheckResourceAttr("hdns_record.bar", "type", "A"),
					resource.TestCheckResourceAttr("hdns_record.bar", "value", "1.1.1.1"),
					resource.TestCheckResourceAttrSet("hdns_record.bar", "zone_id"),
				),
			},
			{
				ResourceName:      "hdns_record.bar",
				ImportState:       true,
				ImportStateVerify: true,
			},
		},
	})
}

func TestAccHdnsRecord_Update(t *testing.T) {
	var record hdns.Record
	rName := acctest.RandomWithPrefix("hdns-testacc")

	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccHdnsPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccHdnsCheckRecordDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccHdnsCheckRecordConfig_create(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccHdnsCheckRecordExists("hdns_record.bar", &record),
					resource.TestCheckResourceAttr("hdns_record.bar", "name", "www"),
					resource.TestCheckResourceAttr("hdns_record.bar", "ttl", "86400"),
					resource.TestCheckResourceAttr("hdns_record.bar", "type", "A"),
					resource.TestCheckResourceAttr("hdns_record.bar", "value", "1.1.1.1"),
					resource.TestCheckResourceAttrSet("hdns_record.bar", "zone_id"),
				),
			},
			{
				Config: testAccHdnsCheckRecordConfig_update(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccHdnsCheckRecordExists("hdns_record.bar", &record),
					resource.TestCheckResourceAttr("hdns_record.bar", "name", "ddd"),
					resource.TestCheckResourceAttr("hdns_record.bar", "ttl", "1800"),
					resource.TestCheckResourceAttr("hdns_record.bar", "type", "A"),
					resource.TestCheckResourceAttr("hdns_record.bar", "value", "2.2.2.2"),
					resource.TestCheckResourceAttrSet("hdns_record.bar", "zone_id"),
				),
			},
		},
	})
}

func TestAccHdnsRecord_MultiRecords(t *testing.T) {
	var record hdns.Record
	rName := acctest.RandomWithPrefix("hdns-testacc")

	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccHdnsPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccHdnsCheckRecordDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccHdnsCheckRecordConfig_multi_records(rName),
				Check: resource.ComposeTestCheckFunc(func() []resource.TestCheckFunc {
					checkFunks := make([]resource.TestCheckFunc, 0)
					for i := 0; i < 10; i++ {
						resourceName := fmt.Sprintf("hdns_record.bar.%d", i)
						checkFunks = append(checkFunks, []resource.TestCheckFunc{
							testAccHdnsCheckRecordExists(resourceName, &record),
							resource.TestCheckResourceAttr(resourceName, "name", fmt.Sprintf("www-%d", i)),
							resource.TestCheckResourceAttr(resourceName, "type", "A"),
							resource.TestCheckResourceAttr(resourceName, "value", fmt.Sprintf("1.1.1.%d", i)),
							resource.TestCheckResourceAttrSet(resourceName, "zone_id"),
						}...)
					}
					return checkFunks
				}()...,
				),
			},
		},
	})
}

func TestAccHdnsRecord_MultiZonesMultiRecords(t *testing.T) {
	//var record hdns.Record
	rName := acctest.RandomWithPrefix("hdns-testacc")

	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccHdnsPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccHdnsCheckRecordDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccHdnsCheckRecordConfig_multi_zones_multi_records(rName),
				Check:  resource.ComposeTestCheckFunc(
				//testAccHdnsCheckRecordExists(`hdns_record.bar["0-0"]`, &record),
				),
			},
		},
	})
}

//noinspection GoSnakeCaseUsage
func testAccHdnsCheckRecordConfig_create(rName string) string {
	return fmt.Sprintf(`
resource "hdns_zone" "foo" {
  name = "%s.dev"
}

resource "hdns_record" "bar" {
  name    = "www"
  type    = "A"
  value   = "1.1.1.1"
  zone_id = hdns_zone.foo.id 
}
`, rName)
}

//noinspection GoSnakeCaseUsage
func testAccHdnsCheckRecordConfig_update(rName string) string {
	return fmt.Sprintf(`
resource "hdns_zone" "foo" {
  name = "%s.dev"
}

resource "hdns_record" "bar" {
  name    = "ddd"
  ttl     = 1800
  type    = "A"
  value   = "2.2.2.2"
  zone_id = hdns_zone.foo.id 
}
`, rName)
}

//noinspection GoSnakeCaseUsage
func testAccHdnsCheckRecordConfig_multi_records(rName string) string {
	return fmt.Sprintf(`
resource "hdns_zone" "foo" {
  name = "%s.dev"
}

resource "hdns_record" "bar" {
  count = 10

  name    = "www-${count.index}"
  type    = "A"
  value   = "1.1.1.${count.index}"
  zone_id = hdns_zone.foo.id
}
`, rName)
}

//noinspection GoSnakeCaseUsage
func testAccHdnsCheckRecordConfig_multi_zones_multi_records(rName string) string {
	return fmt.Sprintf(`
resource "hdns_zone" "foo1" {
  name = "%s-1.dev"
}

resource "hdns_record" "bar1" {
  count = 4

  name    = "www-${count.index}"
  type    = "A"
  value   = "1.1.1.${count.index}"
  zone_id = hdns_zone.foo1.id
}

resource "hdns_zone" "foo2" {
  name = "%s-2.dev"
}

resource "hdns_record" "bar2" {
  count = 4

  name    = "www-${count.index}"
  type    = "A"
  value   = "2.1.1.${count.index}"
  zone_id = hdns_zone.foo2.id
}

resource "hdns_zone" "foo3" {
  name = "%s-3.dev"
}

resource "hdns_record" "bar3" {
  count = 4

  name    = "www-${count.index}"
  type    = "A"
  value   = "3.1.1.${count.index}"
  zone_id = hdns_zone.foo3.id
}
`, rName, rName, rName)
}

func testAccHdnsCheckRecordDestroy(s *terraform.State) error {
	config := testAccProvider.Meta().(*providerConfig)
	client := config.client

	for _, rs := range s.RootModule().Resources {
		if rs.Type != "hdns_record" {
			continue
		}

		record, r, err := client.Record.GetByID(context.Background(), rs.Primary.ID)
		if err != nil && r != nil && r.StatusCode != 404 {
			return fmt.Errorf(
				"error checking if Record (%s) is deleted: %v",
				rs.Primary.ID, err)
		}
		if record != nil {
			return fmt.Errorf("record (%s) has not been deleted", rs.Primary.ID)
		}
	}

	return nil
}

func testAccHdnsCheckRecordExists(r string, record *hdns.Record) resource.TestCheckFunc {
	return func(s *terraform.State) error {
		rs, ok := s.RootModule().Resources[r]

		if !ok {
			return fmt.Errorf("not found: %s", r)
		}

		if rs.Primary.ID == "" {
			return fmt.Errorf("no Record ID is set")
		}

		config := testAccProvider.Meta().(*providerConfig)
		client := config.client

		// Try to find the key
		foundRecord, _, err := client.Record.GetByID(context.Background(), rs.Primary.ID)
		if err != nil {
			return err
		}

		if foundRecord == nil {
			return fmt.Errorf("record not found")
		}

		*record = *foundRecord
		return nil
	}
}


hdns/provider.go
package hdns

import (
	"github.com/alxrem/hdns-go/hdns"
	"github.com/hashicorp/logutils"
	"github.com/hashicorp/terraform-plugin-sdk/helper/logging"
	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
	"github.com/hashicorp/terraform-plugin-sdk/terraform"
	"sync"
)

func Provider() terraform.ResourceProvider {
	return &schema.Provider{
		Schema: map[string]*schema.Schema{
			"token": {
				Type:        schema.TypeString,
				Required:    true,
				DefaultFunc: schema.EnvDefaultFunc("HDNS_TOKEN", nil),
			},
		},
		ResourcesMap: map[string]*schema.Resource{
			"hdns_zone":   resourceZone(),
			"hdns_record": resourceRecord(),
		},
		DataSourcesMap: map[string]*schema.Resource{
			"hdns_zone": dataSourceZone(),
		},
		ConfigureFunc: configureProvider,
	}
}

type providerConfig struct {
	client *hdns.Client
	mu     sync.Map
}

func (pc *providerConfig) Mutex(key interface{}) *sync.Mutex {
	mu, _ := pc.mu.LoadOrStore(key, &sync.Mutex{})
	return mu.(*sync.Mutex)
}

func configureProvider(d *schema.ResourceData) (interface{}, error) {
	opts := []hdns.ClientOption{
		hdns.WithToken(d.Get("token").(string)),
		hdns.WithApplication("hdns-terraform", "0.0.1"),
	}
	if logging.LogLevel() != "" {
		writer, err := logging.LogOutput()
		if err != nil {
			return nil, err
		}
		opts = append(opts, hdns.WithDebugWriter(writer.(*logutils.LevelFilter).Writer))
	}

	client := hdns.NewClient(opts...)

	return &providerConfig{
		client: client,
	}, nil
}


hdns/resource_hdns_zone.go
package hdns

import (
	"context"
	"github.com/alxrem/hdns-go/hdns"
	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
	"log"
)

func resourceZone() *schema.Resource {
	return &schema.Resource{
		Create: resourceZoneCreate,
		Read:   resourceZoneRead,
		Update: resourceZoneUpdate,
		Delete: resourceZoneDelete,
		Importer: &schema.ResourceImporter{
			State: schema.ImportStatePassthrough,
		},

		Schema: map[string]*schema.Schema{
			"name": {
				Type:     schema.TypeString,
				Required: true,
			},
			"ttl": {
				Type:     schema.TypeInt,
				Default:  86400,
				Optional: true,
			},
			"is_secondary_dns": {
				Type:     schema.TypeBool,
				Computed: true,
			},
			"legacy_dns_host": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"legacy_ns": {
				Type:     schema.TypeList,
				Elem:     &schema.Schema{Type: schema.TypeString},
				Computed: true,
			},
			"ns": {
				Type:     schema.TypeList,
				Elem:     &schema.Schema{Type: schema.TypeString},
				Computed: true,
			},
			"owner": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"paused": {
				Type:     schema.TypeBool,
				Computed: true,
			},
			"permission": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"project": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"records_count": {
				Type:     schema.TypeInt,
				Computed: true,
			},
			"registrar": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"status": {
				Type:     schema.TypeString,
				Computed: true,
			},
		},
	}
}

//noinspection GoUnhandledErrorResult
func resourceZoneCreate(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()
	opts := hdns.ZoneCreateOpts{
		Name: d.Get("name").(string),
		TTL:  d.Get("ttl").(int),
	}

	z, _, err := config.client.Zone.Create(ctx, opts)
	if err != nil {
		return err
	}

	d.SetId(z.ID)

	return resourceZoneRead(d, m)
}

//noinspection GoUnhandledErrorResult
func resourceZoneRead(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	id := d.Id()

	z, _, err := config.client.Zone.GetByID(ctx, id)
	if err != nil {
		return err
	}
	if z == nil {
		log.Printf("[WARN] Zone (%s) not found, removing from state", id)
		d.SetId("")
		return nil
	}

	d.SetId(z.ID)
	d.Set("name", z.Name)
	d.Set("ttl", z.TTL)
	d.Set("is_secondary_dns", z.IsSecondaryDNS)
	d.Set("legacy_dns_host", z.LegacyDNSHost)
	d.Set("legacy_ns", z.LegacyNS)
	d.Set("ns", z.NS)
	d.Set("owner", z.Owner)
	d.Set("paused", z.Paused)
	d.Set("permission", z.Permission)
	d.Set("project", z.Project)
	d.Set("records_count", z.RecordsCount)
	d.Set("registrar", z.Registrar)
	d.Set("status", z.Status)

	return nil
}

func resourceZoneUpdate(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	opts := hdns.ZoneUpdateOpts{
		Name: d.Get("name").(string),
		TTL:  d.Get("ttl").(int),
	}

	_, _, err := config.client.Zone.Update(ctx, d.Id(), opts)
	if err != nil {
		if resourceZoneIsNotFound(err, d) {
			return nil
		}
		return err
	}

	return resourceZoneRead(d, m)
}

func resourceZoneDelete(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	id := d.Id()

	r, err := config.client.Zone.Delete(ctx, id)
	if err != nil && r != nil && r.StatusCode != 404 {
		return err
	}

	d.SetId("")

	return nil
}

func resourceZoneIsNotFound(err error, d *schema.ResourceData) bool {
	if hderr, ok := err.(hdns.Error); ok && hderr.Code == hdns.ErrorCodeNotFound {
		log.Printf("[WARN] Zone (%s) not found, removing from state", d.Id())
		d.SetId("")
		return true
	}
	return false
}


hdns/data_source_hdns_zone.go
package hdns

import (
	"context"
	"fmt"
	"github.com/alxrem/hdns-go/hdns"
	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
)

func dataSourceZone() *schema.Resource {
	return &schema.Resource{
		Read: dataSourceZoneRead,
		Schema: map[string]*schema.Schema{
			"id": {
				Type:     schema.TypeString,
				Required: true,
			},
			"name": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"ttl": {
				Type:     schema.TypeInt,
				Computed: true,
			},
			"is_secondary_dns": {
				Type:     schema.TypeBool,
				Computed: true,
			},
			"legacy_dns_host": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"legacy_ns": {
				Type:     schema.TypeList,
				Elem:     &schema.Schema{Type: schema.TypeString},
				Computed: true,
			},
			"ns": {
				Type:     schema.TypeList,
				Elem:     &schema.Schema{Type: schema.TypeString},
				Computed: true,
			},
			"owner": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"paused": {
				Type:     schema.TypeBool,
				Computed: true,
			},
			"permission": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"project": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"records_count": {
				Type:     schema.TypeInt,
				Computed: true,
			},
			"registrar": {
				Type:     schema.TypeString,
				Computed: true,
			},
			"status": {
				Type:     schema.TypeString,
				Computed: true,
			},
		},
	}
}

//noinspection GoUnhandledErrorResult
func dataSourceZoneRead(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	id := d.Get("id").(string)

	z, _, err := config.client.Zone.GetByID(ctx, id)
	if err != nil {
		if hderr, ok := err.(hdns.Error); ok && hderr.Code == hdns.ErrorCodeNotFound {
			return fmt.Errorf("no zone found with id %v", id)
		}
		return err
	}

	d.SetId(z.ID)
	d.Set("name", z.Name)
	d.Set("ttl", z.TTL)
	d.Set("is_secondary_dns", z.IsSecondaryDNS)
	d.Set("legacy_dns_host", z.LegacyDNSHost)
	d.Set("legacy_ns", z.LegacyNS)
	d.Set("ns", z.NS)
	d.Set("owner", z.Owner)
	d.Set("paused", z.Paused)
	d.Set("permission", z.Permission)
	d.Set("project", z.Project)
	d.Set("records_count", z.RecordsCount)
	d.Set("registrar", z.Registrar)
	d.Set("status", z.Status)

	return nil
}


hdns/resource_hdns_record.go
package hdns

import (
	"context"
	"fmt"
	"github.com/alxrem/hdns-go/hdns"
	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
	"log"
)

func resourceRecord() *schema.Resource {
	return &schema.Resource{
		Create: resourceRecordCreate,
		Read:   resourceRecordRead,
		Update: resourceRecordUpdate,
		Delete: resourceRecordDelete,
		Importer: &schema.ResourceImporter{
			State: schema.ImportStatePassthrough,
		},

		Schema: map[string]*schema.Schema{
			"name": {
				Type:     schema.TypeString,
				Required: true,
			},
			"ttl": {
				Type:     schema.TypeInt,
				Default:  86400,
				Optional: true,
			},
			"type": {
				Type:     schema.TypeString,
				Required: true,
			},
			"value": {
				Type:     schema.TypeString,
				Required: true,
			},
			"zone_id": {
				Type:     schema.TypeString,
				Required: true,
			},
		},
	}
}

//noinspection GoUnhandledErrorResult
func resourceRecordCreate(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	zoneID := d.Get("zone_id").(string)
	opts := hdns.RecordCreateOpts{
		Name:   d.Get("name").(string),
		TTL:    d.Get("ttl").(int),
		Type:   d.Get("type").(string),
		Value:  d.Get("value").(string),
		ZoneID: zoneID,
	}

	mu := config.Mutex(zoneID)
	defer func() {
		mu.Unlock()
		log.Printf("[DEBUG] Released lock for zone %s", zoneID)
	}()

	mu.Lock()
	log.Printf("[DEBUG] Acquired lock for zone %s", zoneID)

	record, _, err := config.client.Record.Create(ctx, opts)
	if err != nil {
		return fmt.Errorf("failed to create record \"%s %s\" in the zone %s: %s", opts.Type, opts.Name, opts.ZoneID, err.Error())
	}

	if record, _, err := config.client.Record.GetByID(ctx, record.ID); record == nil && err == nil {
		return fmt.Errorf("failed to create record \"%s %s\" in the zone %s: record is not exists just after creation", opts.Type, opts.Name, opts.ZoneID)
	}

	d.SetId(record.ID)
	d.Set("name", record.Name)
	d.Set("ttl", record.TTL)
	d.Set("type", record.Type)
	d.Set("value", record.Value)
	d.Set("zone_id", record.ZoneID)

	return nil
}

//noinspection GoUnhandledErrorResult
func resourceRecordRead(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	id := d.Id()

	record, _, err := config.client.Record.GetByID(ctx, id)
	if err != nil {
		return fmt.Errorf("failed to read record %s: %s", id, err.Error())
	}
	if record == nil {
		log.Printf("[WARN] Record (%s) not found, removing from state", id)
		d.SetId("")
		return nil
	}

	d.SetId(record.ID)
	d.Set("name", record.Name)
	d.Set("ttl", record.TTL)
	d.Set("type", record.Type)
	d.Set("value", record.Value)
	d.Set("zone_id", record.ZoneID)

	return nil
}

func resourceRecordUpdate(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	zoneID := d.Get("zone_id").(string)
	opts := hdns.RecordUpdateOpts{
		Name:   d.Get("name").(string),
		TTL:    d.Get("ttl").(int),
		Type:   d.Get("type").(string),
		Value:  d.Get("value").(string),
		ZoneID: zoneID,
	}

	mu := config.Mutex(zoneID)
	defer func() {
		mu.Unlock()
		log.Printf("[DEBUG] Released lock for zone %s", zoneID)
	}()

	mu.Lock()
	log.Printf("[DEBUG] Acquired lock for zone %s", zoneID)

	_, _, err := config.client.Record.Update(ctx, d.Id(), opts)
	if err != nil {
		if resourceRecordIsNotFound(err, d) {
			return nil
		}
		return fmt.Errorf("failed to create record \"%s %s\" in the zone %s: %s", opts.Type, opts.Name, opts.ZoneID, err.Error())
	}

	return resourceRecordRead(d, m)
}

func resourceRecordDelete(d *schema.ResourceData, m interface{}) error {
	config := m.(*providerConfig)
	ctx := context.Background()

	id := d.Id()

	zoneID := d.Get("zone_id").(string)
	mu := config.Mutex(zoneID)
	defer func() {
		mu.Unlock()
		log.Printf("[DEBUG] Released lock for zone %s", zoneID)
	}()

	mu.Lock()
	log.Printf("[DEBUG] Acquired lock for zone %s", zoneID)

	r, err := config.client.Record.Delete(ctx, id)
	if err != nil && r != nil && r.StatusCode != 404 {
		return fmt.Errorf("failed to delete record %s: %s", id, err.Error())
	}

	d.SetId("")

	return nil
}

func resourceRecordIsNotFound(err error, d *schema.ResourceData) bool {
	if hderr, ok := err.(hdns.Error); ok && hderr.Code == hdns.ErrorCodeNotFound {
		log.Printf("[WARN] Record (%s) not found, removing from state", d.Id())
		d.SetId("")
		return true
	}
	return false
}
